File: search/src/main/java/net/donggul/api/search/SearchParam.java
package net.donggul.api.search;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;

@JsonIgnoreProperties(ignoreUnknown = true)
public record SearchParam (
        @JsonProperty Integer page,
        @JsonProperty Integer limit,
        @JsonProperty("sort_by") String sortBy,
        @JsonProperty Boolean asc,
        @JsonProperty List<Filter> filters
) {
    public static Object Companion;
    @JsonIgnoreProperties(ignoreUnknown = true)
    public record Filter(
            @JsonProperty String key,
            @JsonProperty String value
    ) { }
}

File: search/src/main/kotlin/net/donggul/api/service/searchParam.kt
package net.donggul.api.service

import com.fasterxml.jackson.databind.ObjectMapper
import net.donggul.api.search.SearchParam
import org.springframework.util.MultiValueMap
import java.util.LinkedList

fun searchParam(om: ObjectMapper, map: MultiValueMap<String, String>): SearchParam = SearchParam(
    map.getFirst("page")?.toIntOrNull(),
    map.getFirst("limit")?.toIntOrNull(),
    map.getFirst("sort_by"),
    map.getFirst("asc")?.toBoolean(),
    if (map.containsKey("filters"))
        om.readValue(
            map.getFirst("filters"),
            om.typeFactory.constructCollectionType(MutableList::class.java, SearchParam.Filter::class.java)
        )
    else LinkedList()
)